text/0000-compiler-requirements.md

```diff
@@ -2,18 +2,45 @@
 
 ## Current State
 
+The entire font compilation pipeline is written in Python, with some parts being written in Cython for speed.
+
 ### Current Tools
 
+fontmake. It's the glue that binds together ufo2ft, fontTools and some other codebases to take any of the usual font formats (UFOs, Designspace + UFOs and Glyphs.app files) and produce font binaries (TTF, CFF(2); static, interpolated static, variable).
+
 ### Key Limitations
 
+* Python is bad at embarrasingly parallel tasks like instance generation and glyph compilation due to overhead and lacking multicore facilities.
```

Comment:
even if we imagine a ninja'fied fontmake?

Reply:
Instance generation works fine, which is (partly) what I'm doing with gftools.builder.ninja. A gftools-builder run (which is intended to bring an entire family to production) is a much larger superset of operations than a fontmake run:

* It typically runs on more than one source file at once (i.e. Upright and Italic sources for a family)
* It typically builds a wide variety of output fonts (variable TTF, TTF, OTF)
* There are typically other steps that you want to orchestrate in a `make`-like context (autohinting each TTF output, converting each TTF to a webfont, running hotfix scripts, converting each TTF output to building STAT tables when all output variable TTFs are complete)

So that gains a lot from using ninja. I get roughly a 3x speedup.

It is possible to break up some of the operations that fontmake does and use ninja to orchestrate them, in *some* contexts. Generally this is when you are dealing with static instances (`-i` flag):

* Creating instance UFOs from a Designspace file is currently done in serial, can be done in parallel.
* Creating instance binaries from each instance UFO is current done in serial, can be done in parallel.
* Creating master binaries from each master UFO is current done in serial, can be done in parallel.

On the other hand, building a single variable font would only involve the last of these operations, and I'm not convinced that's enough of a gain to warrant the work of adding ninja orchestration.

Just so you see the degree of separable operations in a gftools-builder run, here is a `ninja -t graph` dump of `gftools.builder.ninja` for a typical font family:

![ninja](https://user-images.githubusercontent.com/106728/172669662-e6f567c7-33ca-455a-a647-224e14d2f7d1.png)